src/libmongoc/src/mongoc/mongoc-scram.c

```diff
@@ -1052,81 +1052,160 @@ _mongoc_sasl_prep_impl (const char *name,
+   /* 1. convert str to Unicode codepoints. */
    /* preflight to get the destination length. */
-   (void) u_strFromUTF8 (
-      NULL, 0, &in_utf16_len, in_utf8, in_utf8_len, &error_code);
-   if (error_code != U_BUFFER_OVERFLOW_ERROR) {
-      SASL_PREP_ERR_RETURN ("could not calculate UTF-16 length of %s");
-   }
-
-   /* convert to UTF-16. */
-   error_code = U_ZERO_ERROR;
-   in_utf16 = bson_malloc (sizeof (UChar) *
-                           (in_utf16_len + 1)); /* add one for null byte. */
-   (void) u_strFromUTF8 (
-      in_utf16, in_utf16_len + 1, NULL, in_utf8, in_utf8_len, &error_code);
-   if (error_code) {
-      bson_free (in_utf16);
-      SASL_PREP_ERR_RETURN ("could not convert %s to UTF-16");
-   }
-
-   /* 2. perform SASLPrep. */
-   prep = usprep_openByType (USPREP_RFC4013_SASLPREP, &error_code);
-   if (error_code) {
-      bson_free (in_utf16);
-      SASL_PREP_ERR_RETURN ("could not start SASLPrep for %s");
-   }
-   /* preflight. */
-   out_utf16_len = usprep_prepare (
-      prep, in_utf16, in_utf16_len, NULL, 0, USPREP_DEFAULT, NULL, &error_code);
-   if (error_code != U_BUFFER_OVERFLOW_ERROR) {
-      bson_free (in_utf16);
-      usprep_close (prep);
-      SASL_PREP_ERR_RETURN ("could not calculate SASLPrep length of %s");
-   }
-
-   /* convert. */
-   error_code = U_ZERO_ERROR;
-   out_utf16 = bson_malloc (sizeof (UChar) * (out_utf16_len + 1));
-   (void) usprep_prepare (prep,
-                          in_utf16,
-                          in_utf16_len,
-                          out_utf16,
-                          out_utf16_len + 1,
-                          USPREP_DEFAULT,
-                          NULL,
-                          &error_code);
-   if (error_code) {
-      bson_free (in_utf16);
-      bson_free (out_utf16);
-      usprep_close (prep);
-      SASL_PREP_ERR_RETURN ("could not execute SASLPrep for %s");
-   }
-   bson_free (in_utf16);
-   usprep_close (prep);
-
-   /* 3. convert back to UTF-8. */
-   /* preflight. */
-   (void) u_strToUTF8 (
-      NULL, 0, &out_utf8_len, out_utf16, out_utf16_len, &error_code);
-   if (error_code != U_BUFFER_OVERFLOW_ERROR) {
-      bson_free (out_utf16);
+   num_chars = _mongoc_utf8_string_length (in_utf8);
+   if (num_chars == -1) {
       SASL_PREP_ERR_RETURN ("could not calculate UTF-8 length of %s");
    }
 
-   /* convert. */
-   error_code = U_ZERO_ERROR;
-   out_utf8 = (char *) bson_malloc (
-      sizeof (char) * (out_utf8_len + 1)); /* add one for null byte. */
-   (void) u_strToUTF8 (
-      out_utf8, out_utf8_len + 1, NULL, out_utf16, out_utf16_len, &error_code);
-   if (error_code) {
-      bson_free (out_utf8);
-      bson_free (out_utf16);
-      SASL_PREP_ERR_RETURN ("could not convert %s back to UTF-8");
+   /* convert to unicode. */
+   utf8_codepoints = bson_malloc (sizeof (unsigned int) *
+                                  (num_chars + 1)); /* add one for null byte. */
+   c = in_utf8;
+   for (i = 0; i < num_chars; ++i) {
```

Comment:
```suggestion
   for (size_t i = 0; i < num_chars; ++i) {
```
C99 permits mixing declarations and code. Recommend reducing scope of local variables.

Reply:
I removed all of the variable declarations from the top of the function and only kept the variables that I determined weren't "local" to a part of the function. I recognize how that is subjective though so let me know if I should remove more.